## Part 1: Preprocessing

In [94]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [95]:
# Determine the number of unique values in each column
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [96]:
# Create y_df with the Attrition and Department columns
y_df = pd.DataFrame(attrition_df, columns=['Attrition', 'Department'])
y_df.head()

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [97]:
# Create a list of at least 10 column names to use as X data

selected_columns = ['Age', 'HourlyRate', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'PercentSalaryHike', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'YearsSinceLastPromotion']

# Create X_df using your selected columns
X_df = attrition_df[selected_columns]

# Show the data types for X_df
X_df.dtypes

Age                        int64
HourlyRate                 int64
DistanceFromHome           int64
Education                  int64
EnvironmentSatisfaction    int64
PercentSalaryHike          int64
JobInvolvement             int64
JobLevel                   int64
JobSatisfaction            int64
YearsSinceLastPromotion    int64
dtype: object

In [98]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)

In [99]:
# Convert your X data to numeric data types however you see fit

#convert categorical columns to numeric
X_numeric = pd.get_dummies(X_df)

#Display first rowns of transformed data
X_numeric.head()



Age  HourlyRate  DistanceFromHome  Education  EnvironmentSatisfaction  \
0   41          94                 1          2                        2   
1   49          61                 8          1                        3   
2   37          92                 2          2                        4   
3   33          56                 3          4                        4   
4   27          40                 2          1                        1   

   PercentSalaryHike  JobInvolvement  JobLevel  JobSatisfaction  \
0                 11               3         2                4   
1                 23               2         2                2   
2                 15               2         1                3   
3                 11               3         1                3   
4                 12               3         1                2   

   YearsSinceLastPromotion  
0                        0  
1                        1  
2                        0  
3                        3  
4                        2

In [100]:
#identify numeric missing values
any_missing = X_df.isna().any().any()
print(any_missing)

False


In [101]:

# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_train_scaled = scaler.fit(X_train)

# Scale the training and testing data
X_test_scaled = scaler.transform(X_test)
X_train_scaled = scaler.transform(X_train)

In [102]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

# Create a OneHotEncoder for the Department column
encoder = OneHotEncoder(sparse_output=False)


# Fit the encoder to the training data
y_department_train_encoded = encoder.fit_transform(np.array(y_train['Department']).reshape(-1, 1))



# Create two new variables by applying the encoder to the training and testing data
y_department_test_encoded = encoder.transform(np.array(y_test['Department']).reshape(-1, 1))


print(y_department_train_encoded)


[[0. 1. 0.]
 [0. 1. 0.]
 [0. 0. 1.]
 ...
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [103]:
from sklearn.preprocessing import OneHotEncoder
import numpy as np

# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
attrition_train_encoded = attrition_encoder.fit(np.array(y_train['Attrition']).reshape(-1, 1))


# Create two new variables by applying the encoder to the training and testing data
y_train_attrition_encoded = attrition_encoder.transform(np.array(y_train['Attrition']).reshape(-1, 1))
y_test_attrition_encoded = attrition_encoder.transform(np.array(y_test['Attrition']).reshape(-1, 1))

print(y_train_attrition_encoded)


[[1. 0.]
 [1. 0.]
 [1. 0.]
 ...
 [0. 1.]
 [1. 0.]
 [1. 0.]]


## Part 2: Create, Compile, and Train the Model

In [104]:
# Find the number of unique categories in the Department column
unique_departments = y_train['Department'].unique()
print(f"Unique categories in Department: {unique_departments}")

Unique categories in Department: ['Research & Development' 'Sales' 'Human Resources']


In [105]:
from tensorflow.keras.layers import Input, Dense

# Find the number of columns in the X training data
input_dim = X_train_scaled.shape[1]

# Create the input layer
input_layer = layers.Input(shape=(input_dim,))

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(32, activation='relu')(shared_layer1)

In [106]:


# Create a branch for Department with a hidden layer and an output layer


# Create the hidden layer
department_hidden_layer = layers.Dense(16, activation='relu')(shared_layer2)



# Create the output layer
department_output_layer = layers.Dense(y_department_train_encoded.shape[1], activation='sigmoid', name='department_output')(department_hidden_layer)


In [107]:
from tensorflow.keras.layers import Dense

# Create a branch for Attrition with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden_layer = Dense(16, activation='relu')(shared_layer2)

# Create the output layer
attrition_output_layer = layers.Dense(y_train_attrition_encoded.shape[1], activation='sigmoid', name='attrition_output')(attrition_hidden_layer)

In [108]:
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam

# Create the model
model = Model(
    inputs=input_layer,  # Input layer
    outputs=[department_output_layer, attrition_output_layer]
)

# Compile the model
model.compile(
    optimizer=Adam(learning_rate=0.001),  # Optimizer
    loss={
        'department_output': 'categorical_crossentropy',  # Loss for Department branch
        'attrition_output': 'binary_crossentropy'  # Loss for Attrition branch
    },
    metrics={
        'department_output': 'accuracy',  # Metric for Department branch
        'attrition_output': 'accuracy'  # Metric for Attrition branch
    }
)


# Summarize the model
model.summary()

Model: "functional_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer_5             │ (None, 10)             │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_20 (Dense)          │ (None, 64)             │            704 │ input_layer_5[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_21 (Dense)          │ (None, 32)             │          2,080 │ dense_20[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_22 (Dense)          │ (None, 16)             │            528 │ dense_21[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_23 (Dense)          │ (None, 16)             │            528 │ dense_21[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             51 │ dense_22[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             34 │ dense_23[0][0]         │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 3,925 (15.33 KB)

 Trainable params: 3,925 (15.33 KB)

 Non-trainable params: 0 (0.00 B)

In [109]:

# Train the model
history = model.fit(
    X_train_scaled,  # Input features
    {
        'department_output': y_department_train_encoded,
        'attrition_output': y_train_attrition_encoded
    },
    validation_split=0.2,
    epochs=50,
    verbose=1
)

Epoch 1/50
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 20ms/step - attrition_output_accuracy: 0.5310 - attrition_output_loss: 0.6948 - department_output_accuracy: 0.4296 - department_output_loss: 1.0499 - loss: 1.7448 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.6180 - val_department_output_accuracy: 0.6314 - val_department_output_loss: 0.8819 - val_loss: 1.5121
Epoch 2/50
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - attrition_output_accuracy: 0.8266 - attrition_output_loss: 0.5851 - department_output_accuracy: 0.6460 - department_output_loss: 0.8349 - loss: 1.4204 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.5276 - val_department_output_accuracy: 0.6314 - val_department_output_loss: 0.8146 - val_loss: 1.3680
Epoch 3/50
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - attrition_output_accuracy: 0.8261 - attrition_output_loss: 0.4911 - department_output_accuracy: 0.6411 - department_output_loss: 0.7890 - loss: 1.2801 - val_attrition_output_accuracy: 0.7966 - va

In [110]:
# Evaluate the model with the testing data
test_loss, test_attrition_loss, test_department_loss, test_attrition_accuracy, test_department_accuracy = model.evaluate(
    X_test_scaled,
    {
        'department_output': y_department_test_encoded,
        'attrition_output': y_test_attrition_encoded
    }
)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.8310 - attrition_output_loss: 0.4268 - department_output_accuracy: 0.6278 - department_output_loss: 0.8406 - loss: 1.2764 


In [111]:
# Print the accuracy for both department and attrition
print(f"Test Accuracy (Department): {test_department_accuracy}")
print(f"Test Accuracy (Attrition): {test_attrition_accuracy}")

Test Accuracy (Department): 0.6598639488220215
Test Accuracy (Attrition): 0.8367347121238708


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy appears to be the best metric due to it's high accuracy, showing the model test accuracy for attrition at 83.67% and the output accuracy for attrition at 83.10%. However, it is worth noting that the test accuracy for department comes in at 65.99%, and the output accuracy for department comes in at 62.78%.
2. I used Relu for activation of the hidden layers for both Attrition and Department. I used Sigmoid for the output of both Attrition and Deparment layers because of how they complement each other for binary classification.
3. The model could be improved by re-evaluating the epochs and looking at other types of activation. Looking at precision and recall may also help.